slsDetectorSoftware/src/Module.cpp

```diff
@@ -1746,19 +1747,109 @@ void Module::setLEDEnable(bool enable) {
+                                       ToString(args));
+                }
+                pat.patlimits[0] = StringTo<uint32_t>(args[1]);
+                pat.patlimits[1] = StringTo<uint32_t>(args[2]);
+                if (pat.patlimits[0] >= MAX_PATTERN_LENGTH ||
+                    pat.patlimits[1] >= MAX_PATTERN_LENGTH) {
+                    throw RuntimeError("Invalid address for " + ToString(args));
+                }
+            } else if (cmd == "patloop0" || cmd == "patloop1" ||
+                       cmd == "patloop2") {
+                if (nargs != 2) {
+                    throw RuntimeError("Invalid arguments for " +
+                                       ToString(args));
+                }
+                int level = cmd[cmd.find_first_of("012")] - '0';
+                int patloop1 = StringTo<uint32_t>(args[1]);
+                int patloop2 = StringTo<uint32_t>(args[2]);
+                pat.patloop[level * 2 + 0] = patloop1;
+                pat.patloop[level * 2 + 1] = patloop2;
+                if (patloop1 >= MAX_PATTERN_LENGTH ||
+                    patloop2 >= MAX_PATTERN_LENGTH) {
+                    throw RuntimeError("Invalid address for " + ToString(args));
+                }
+            } else if (cmd == "patnloop0" || cmd == "patnloop1" ||
+                       cmd == "patnloop2") {
+                if (nargs != 1) {
+                    throw RuntimeError("Invalid arguments for " +
+                                       ToString(args));
+                }
+                int level = cmd[cmd.find_first_of("012")] - '0';
+                pat.patnloop[level] = StringTo<uint32_t>(args[1]);
+            } else if (cmd == "patwait0" || cmd == "patwait1" ||
+                       cmd == "patwait2") {
+                if (nargs != 1) {
+                    throw RuntimeError("Invalid arguments for " +
+                                       ToString(args));
+                }
+                int level = cmd[cmd.find_first_of("012")] - '0';
+                pat.patwait[level] = StringTo<uint32_t>(args[1]);
+                if (pat.patwait[level] >= MAX_PATTERN_LENGTH) {
+                    throw RuntimeError("Invalid address for " + ToString(args));
+                }
+            } else if (cmd == "patwaittime0" || cmd == "patwaittime1" ||
+                       cmd == "patwaittime2") {
+                if (nargs != 1) {
+                    throw RuntimeError("Invalid arguments for " +
+                                       ToString(args));
+                }
+                int level = cmd[cmd.find_first_of("012")] - '0';
+                pat.patwaittime[level] = StringTo<uint64_t>(args[1]);
+            } else {
+                throw RuntimeError("Unknown command in pattern file " + cmd);
+            }
         }
-        fclose(fd);
-    } else {
-        throw RuntimeError("Could not open file to set pattern");
     }
+    input_file.close();
```

Comment:
The file will be closed by the destructor of input_file once leaving the scope

Reply:
done